File: src/tests/tiff.zig
const std = @import("std");
const generateGradientPixels = @import("shared.zig").generateGradientPixels;

test "tiff encode" {
    const c = @cImport({
        @cInclude("tiffio.h");
    });

    const width: u32 = 16;
    const height: u32 = 16;

    const pixels = try generateGradientPixels(std.testing.allocator, width, height, true);
    defer std.testing.allocator.free(pixels);

    var temp_dir = std.testing.tmpDir(.{});
    defer temp_dir.cleanup();
    (try temp_dir.dir.createFile("temp_output.tiff", .{})).close();
    var output_path_buf = std.mem.zeroes([std.fs.max_path_bytes]u8);
    const output_path = try temp_dir.dir.realpathZ("temp_output.tiff", &output_path_buf);

    const tif = c.TIFFOpen(output_path.ptr, "w") orelse return error.FailedToCreateTIFF;
    defer _ = c.TIFFClose(tif);

    _ = c.TIFFSetField(tif, c.TIFFTAG_IMAGEWIDTH, @as(c_uint, width));
    _ = c.TIFFSetField(tif, c.TIFFTAG_IMAGELENGTH, @as(c_uint, height));
    _ = c.TIFFSetField(tif, c.TIFFTAG_SAMPLESPERPIXEL, @as(c_int, 4));
    _ = c.TIFFSetField(tif, c.TIFFTAG_BITSPERSAMPLE, @as(c_int, 8));
    _ = c.TIFFSetField(tif, c.TIFFTAG_ORIENTATION, @as(c_int, c.ORIENTATION_TOPLEFT));
    _ = c.TIFFSetField(tif, c.TIFFTAG_PLANARCONFIG, @as(c_int, c.PLANARCONFIG_CONTIG));
    _ = c.TIFFSetField(tif, c.TIFFTAG_PHOTOMETRIC, @as(c_int, c.PHOTOMETRIC_RGB));
    _ = c.TIFFSetField(tif, c.TIFFTAG_COMPRESSION, @as(c_int, c.COMPRESSION_NONE));
    _ = c.TIFFSetField(tif, c.TIFFTAG_ROWSPERSTRIP, @as(c_uint, height));

    if (c.TIFFWriteEncodedStrip(tif, @as(u32, 0), pixels.ptr, @as(c_long, @intCast(pixels.len))) == -1) {
        return error.FailedToEncode;
    }

    const stat = try temp_dir.dir.statFile("temp_output.tiff");
    try std.testing.expect(stat.size > 0);
}

test "tiff decode" {
    const c = @cImport({
        @cInclude("tiffio.h");
    });

    const image_path = "src/tests/test-images/orange.tiff";
    const tif = c.TIFFOpen(image_path, "r") orelse return error.FailedToOpenTIFF;
    defer _ = c.TIFFClose(tif);

    var width: u32 = 0;
    var height: u32 = 0;
    _ = c.TIFFGetField(tif, c.TIFFTAG_IMAGEWIDTH, &width);
    _ = c.TIFFGetField(tif, c.TIFFTAG_IMAGELENGTH, &height);

    const pixels = try std.testing.allocator.alloc(u32, @as(usize, width) * @as(usize, height));
    defer std.testing.allocator.free(pixels);

    if (c.TIFFReadRGBAImage(tif, width, height, pixels.ptr, 0) == 0) {
        return error.FailedToDecode;
    }

    // Verify that we have some pixels
    try std.testing.expect(pixels.len > 0);
}

File: src/tests/tiff.c
#include <stdlib.h>
#include <tiffio.h>

#include "shared.h"

bool test_tiff_encode() {
  const size_t width = 16;
  const size_t height = 16;

  auto pixels = generate_gradient_pixels(width, height, true);

  const char *output_path = "./zig-cache/test-outputs/tiff_encode.tiff";
  TIFF *tif = TIFFOpen(output_path, "w");
  if (!tif) {
    free(pixels.data);
    return false;
  }

  TIFFSetField(tif, TIFFTAG_IMAGEWIDTH, (uint32_t)width);
  TIFFSetField(tif, TIFFTAG_IMAGELENGTH, (uint32_t)height);
  TIFFSetField(tif, TIFFTAG_SAMPLESPERPIXEL, 4);
  TIFFSetField(tif, TIFFTAG_BITSPERSAMPLE, 8);
  TIFFSetField(tif, TIFFTAG_ORIENTATION, ORIENTATION_TOPLEFT);
  TIFFSetField(tif, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
  TIFFSetField(tif, TIFFTAG_PHOTOMETRIC, PHOTOMETRIC_RGB);
  TIFFSetField(tif, TIFFTAG_COMPRESSION, COMPRESSION_NONE);
  TIFFSetField(tif, TIFFTAG_ROWSPERSTRIP, (uint32_t)height);

  if (TIFFWriteEncodedStrip(tif, 0, pixels.data, pixels.len) == -1) {
    free(pixels.data);
    TIFFClose(tif);
    return false;
  }

  free(pixels.data);
  TIFFClose(tif);

  // Check if file exists and has size > 0
  FILE *f = fopen(output_path, "rb");
  if (!f)
    return false;
  fseek(f, 0, SEEK_END);
  size_t size = ftell(f);
  fclose(f);
  remove(output_path);
  return size > 0;
}

bool test_tiff_decode() {
  const char *image_path = "src/tests/test-images/orange.tiff";
  TIFF *tif = TIFFOpen(image_path, "r");
  if (!tif)
    return false;

  uint32_t width, height;
  TIFFGetField(tif, TIFFTAG_IMAGEWIDTH, &width);
  TIFFGetField(tif, TIFFTAG_IMAGELENGTH, &height);

  uint32_t *pixels = malloc(width * height * sizeof(uint32_t));
  if (!pixels) {
    TIFFClose(tif);
    return false;
  }

  if (TIFFReadRGBAImage(tif, width, height, pixels, 0) == 0) {
    free(pixels);
    TIFFClose(tif);
    return false;
  }

  free(pixels);
  TIFFClose(tif);
  return true;
}

int main() {
  if (!test_tiff_encode()) {
    printf("tiff_encode failed\n");
    return 1;
  }

  if (!test_tiff_decode()) {
    printf("tiff_decode failed\n");
    return 1;
  }

  return 0;
}
